spec/android_unit_test_pr_checker_spec.rb

```diff
@@ -0,0 +1,222 @@
+          'project/src/androidTest/java/org/test/ToolTest.kt' => 'class ToolTest { fun testMethod() {} }',
+          'File2.kt' => 'class Abcdef(name: String) { public fun testMe() { println(""); } }'
+        }
+
+        bypass_label = 'ignore-no-tests'
+
+        diff = generate_add_diff(changes_dict)
+        allow(@dangerfile.git).to receive(:diff).and_return(diff)
+        allow(@plugin.github).to receive(:pr_labels).and_return([bypass_label])
+
+        @plugin.check_missing_tests(bypass_label: bypass_label)
+
+        expect(@dangerfile.status_report[:errors]).to be_empty
+      end
+
+      it 'does not show that a PR adding custom classes / subclasses patterns are missing tests' do
+        exceptions = [
+          /ViewHelper$/
+        ].freeze
+
+        subclasses_exceptions = [
+          /BaseViewWrangler/
+        ].freeze
+
+        changes_dict = {
+          'File1.java' => 'import java.utils.*;\n\n public class Abc extends BaseViewWrangler { public static void main(String[] args) { println(""); } }',
+          'AbcWrangler.java' => 'import java.utils.*;\n\n abstract class AbcWrangler extends BaseViewWrangler { public abstract void wrangle(); }',
+          'KotlinWrangler.kt' => 'abstract class KotlinWrangler: BaseViewWrangler { abstract fun wrangle(); }',
+          'project/src/androidTest/java/org/test/ToolTest.kt' => 'class ToolTest { void testMethod() {} }',
+          'AbcdefViewHelper.kt' => 'class AbcdefViewHelper(name: String) { fun testMe() { println(""); } }',
+          'AbcdefgViewHelper.java' => 'public final class AbcdefgViewHelper { public static void testMe() { System.out.println(""); } }'
+        }
+
+        diff = generate_add_diff(changes_dict)
+        allow(@dangerfile.git).to receive(:diff).and_return(diff)
+
+        @plugin.check_missing_tests(classes_exceptions: exceptions, subclasses_exceptions: subclasses_exceptions)
+
+        expect(@dangerfile.status_report[:errors]).to be_empty
+      end
+    end
+
+    def generate_add_diff(changes_dict)
+      changes_dict.map do |file_path, content|
+        diff_str = <<~PATCH
+          diff --git a/#{file_path} b/#{file_path}
+          new file mode 100644
+          index 0000000..fd48a22
+          --- /dev/null
+          +++ b/#{file_path}
+          @@ -0,0 +1 @@
+          +#{content}
+          \\ No newline at end of file
+        PATCH
+
+        OpenStruct.new(type: 'new', path: file_path, patch: diff_str)
+      end
+    end
+
+    def generate_deleted_diff(changes_dict)
+      changes_dict.map do |file_path, content|
+        diff_str = <<~PATCH
+          diff --git a/#{file_path} b/#{file_path}
+          deleted file mode 100644
+          index fd48a22..0000000
+          --- a/#{file_path}
+          +++ /dev/null
+          @@ -1 +0,0 @@
+          -#{content}
+          \\ No newline at end of file
+        PATCH
+
+        OpenStruct.new(type: 'deleted', path: file_path, patch: diff_str)
+      end
+    end
```

Comment:
That just made me think: we probably want to also test the case of a diff that contains a mix of both added and removed lines inside the same file.

Typical example is for checking that if a class definition, or a test using that class, moved within a file, we will not incorrectly consider it an addition or a deletion, but correctly [`reject!` it](https://github.com/Automattic/dangermattic/pull/1/files#diff-7006d15ff302c846b3271adb5945d3fd455d7c13d35fca97b46444e9f3ede9e4R108) from the violations instead.

e.g. say we just moved the `class Foo` implementation some place later in the `Foo.kt` file…
```
--- a/Foo.kt
+++ b/Foo.kt
@@ -3 +0,0 @@
- class Foo {
-   fun demo() { }
- }
@@ -27,0 +3 @@
- class Foo {
-   fun demo() { }
- }
```

Reply:
Good idea! Just pushed some changes and a new test adding a similar diff 👍